File: core/core-api/src/main/java/com/coffee/gu/coupon/CouponManager.java
package com.coffee.gu.coupon;

import com.coffee.gu.CoreException;
import com.coffee.gu.ErrorType;
import com.coffee.gu.Principal;
import com.coffee.gu.enums.IssuedCouponState;
import com.coffee.gu.enums.PrincipalType;

import org.springframework.stereotype.Component;

@Component
public class CouponManager {

    private final IssuedCouponRepository issuedCouponRepository;

    public CouponManager(IssuedCouponRepository issuedCouponRepository) {
        this.issuedCouponRepository = issuedCouponRepository;
    }

    public void issue(Principal principal, Coupon coupon) {
        if (principal.getType() == PrincipalType.GUEST) throw new CoreException(ErrorType.UNAUTHORIZED, null);
        issuedCouponRepository.save(IssuedCoupon.download(principal, coupon));
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/coupon/CouponManager.kt
package com.coffee.gu.coupon

import com.coffee.gu.CoreException
import com.coffee.gu.ErrorType
import com.coffee.gu.Principal
import com.coffee.gu.enums.PrincipalType
import org.springframework.stereotype.Component

@Component
class CouponManager(
    private val issuedCouponRepository: IssuedCouponRepository
) {
    fun issue(principal: Principal, coupon: Coupon) {
        if (principal.type == PrincipalType.GUEST) throw CoreException(ErrorType.UNAUTHORIZED, null)
        issuedCouponRepository.save(IssuedCoupon.download(principal, coupon))
    }
}
